### PR TITLE
Add top_p, top_k parameters, add defaults to all fields

### LINKED ISSUE #236: Support for topP in BasePromptDriver
**Is your feature request related to a problem? Please describe.**
[Top P](https://docs.cohere.com/docs/controlling-generation-with-top-k-top-p) is a common parameter to be passed to LLM when generating a completion of a prompt.

**Describe the solution you'd like**
I would like to have top P as part of the base model driver

**Describe alternatives you've considered**
I could add Top P as a specific parameter of my model driver, but that would not allow me to pass the Top P value to existing model drivers

**Additional context**

### LINKED ISSUE #293: Evaluate which other configuration variables are missing (e.g. Top K)

### PR DESCRIPTION
**Pull request by collindutter**
- [x] I have read and agree to the contributing guidelines for [submitting new pull requests](https://github.com/griptape-ai/griptape?tab=readme-ov-file#submitting-pull-requests).

## Describe your changes
- Added type hints to Bedrock clients.
- Reorganized Driver fields to place dependent fields (clients/tokenizers) after primitive fields.
- Added `top_p` field to all Prompt Drivers.
- Added `BasePromptDriver.additional_params` to set additional request parameters such as top k.
- Added defaults to all fields. This solves an issue where users would provide a `client` but still be required to pass in fields like `api_key`.
- Removed redundant `model` fields as it is provided by the base class.
## Issue ticket number and link
Closes #236
Closes #293

### CODE REVIEW

**Comment by collindutter** [comment]
@dylanholmes @vasinov need some feedback on this. What's the criteria for adding a field to a Prompt Driver? For instance, do we include top_k on the Prompt Drivers that support it? What about more niche parameters? 

It feels unrealistic to support all possible parameters which is why I added `additional_params`. But which fields should live on the Prompt Driver vs relying on the user to provide it via `additional_params`?

**Comment by dylanholmes** [comment]
> @dylanholmes @vasinov need some feedback on this. What's the criteria for adding a field to a Prompt Driver? For instance, do we include top_k on the Prompt Drivers that support it? What about more niche parameters?
> 
> It feels unrealistic to support all possible parameters which is why I added `additional_params`. But which fields should live on the Prompt Driver vs relying on the user to provide it via `additional_params`?

I think that we should provide translation for all params that are supported by at least 2 drivers and someone explicitly asks for it to be bumped up to `BasePromptDriver`.

Options I can think of are:
1. The way done in this PR (common params as optional properties on `BasePromptDriver`)
2. A separate class for `PromptDriverParams` that just has the param related properties from `BasePromptDriver`
3. Put all the params in a single dict in `BasePromptDriver`, but perform translation for some specific keys (like `top_p` -> `topP` in bedrock for example)

2 would be preferable over 1 if the list of params was greater than say 5 imo. 3 doesn't seem like it'd be error prone, so I don't like it.

I think maybe we just start with 1 and move to 2 if needed some time in the future.

File: griptape/drivers/prompt/amazon_bedrock_prompt_driver.py [hunk 1]
```diff
@@ -98,8 +105,11 @@ def _base_params(self, prompt_stack: PromptStack) -> dict:
             "modelId": self.model,
             "messages": messages,
             "system": system_messages,
-            "inferenceConfig": {"temperature": self.temperature},
-            "additionalModelRequestFields": self.additional_model_request_fields,
+            "inferenceConfig": {
+                **({"temperature": self.temperature} if self.temperature is not None else {}),
+                **({"topP": self.top_p} if self.top_p is not None else {}),
+            },
+            "additionalModelRequestFields": self.additional_params,
```
Comment (dylanholmes) [inline griptape/drivers/prompt/amazon_bedrock_prompt_driver.py:112]: This seems awkward. Would anyone ever need to pass additional params to `inferenceConfig`? (Like wouldn't top k go there?)

**Review by dylanholmes — CHANGES_REQUESTED** [review]
Looking forward to ironing out these params!

File: griptape/drivers/prompt/huggingface_hub_prompt_driver.py [hunk 2]
```diff
@@ -28,10 +28,9 @@ class HuggingFaceHubPromptDriver(BasePromptDriver):
         tokenizer: Custom `HuggingFaceTokenizer`.
     """
 
-    api_token: str = field(kw_only=True, metadata={"serializable": True})
+    api_token: Optional[str] = field(kw_only=True, default=None, metadata={"serializable": True})
     max_tokens: int = field(default=250, kw_only=True, metadata={"serializable": True})
-    params: dict = field(factory=dict, kw_only=True, metadata={"serializable": True})
-    model: str = field(kw_only=True, metadata={"serializable": True})
+    top_p: Optional[float] = field(default=None, kw_only=True, metadata={"serializable": True})
```
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/prompt/huggingface_hub_prompt_driver.py:33]: Oops, fixed.

File: griptape/drivers/prompt/huggingface_pipeline_prompt_driver.py [hunk 3]
```diff
@@ -26,8 +26,7 @@ class HuggingFacePipelinePromptDriver(BasePromptDriver):
     """
 
     max_tokens: int = field(default=250, kw_only=True, metadata={"serializable": True})
-    model: str = field(kw_only=True, metadata={"serializable": True})
-    params: dict = field(factory=dict, kw_only=True, metadata={"serializable": True})
+    top_p: Optional[float] = field(default=None, kw_only=True, metadata={"serializable": True})
```
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/prompt/huggingface_pipeline_prompt_driver.py]: Oops, fixed.

File: griptape/drivers/prompt/openai_chat_prompt_driver.py [hunk 4]
```diff
@@ -55,26 +55,14 @@ class OpenAiChatPromptDriver(BasePromptDriver):
...
     user: str = field(default="", kw_only=True, metadata={"serializable": True})
     response_format: Optional[Literal["json_object"]] = field(
         default=None,
         kw_only=True,
         metadata={"serializable": True},
     )
     seed: Optional[int] = field(default=None, kw_only=True, metadata={"serializable": True})
-    tool_choice: str = field(default="auto", kw_only=True, metadata={"serializable": False})
-    use_native_tools: bool = field(default=True, kw_only=True, metadata={"serializable": True})
+    top_p: Optional[float] = field(default=None, kw_only=True, metadata={"serializable": True})
```
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/prompt/openai_chat_prompt_driver.py]: Oops, fixed.

File: griptape/drivers/prompt/base_prompt_driver.py [hunk 5]
```diff
@@ -42,14 +42,16 @@ class BasePromptDriver(SerializableMixin, ExponentialBackoffMixin, ABC):
         use_native_tools: Whether to use LLM's native function calling capabilities. Must be supported by the model.
     """
 
-    temperature: float = field(default=0.1, metadata={"serializable": True})
+    temperature: Optional[float] = field(default=0.1, metadata={"serializable": True})
+    top_p: Optional[float] = field(default=None, kw_only=True, metadata={"serializable": True})
```
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/prompt/base_prompt_driver.py:46]: Good question. In some places we use an `attrs.validator`, and in others we override the field and set `init=False` so the field doesn't even appear as available to the user.

File: griptape/drivers/prompt/amazon_bedrock_prompt_driver.py (hunk 1 above)
Comment (collindutter) [reply griptape/drivers/prompt/amazon_bedrock_prompt_driver.py:112]: Ah good point -- maybe we can take inspiration from how Bedrock has split out their fields here? 
`inferenceConfig` = top level Driver fields -- common to all models.
`additionalmodelRequestFields` = `additional/native_params` -- model specific parameters.

**Comment by collindutter** [comment]
@dylanholmes @vasinov another possible solution here is to get clever with attrs by allowing users to pass in arbitrary fields beyond what we've defined. Something like this:

```python
    def __init__(self, *args, **kwargs) -> None:
        self._params = {}

        keys_to_delete = []
        attrs_fields = fields_dict(self.__class__)

        for key, value in kwargs.items():
            if key not in attrs_fields:
                self._params[key] = value
                keys_to_delete.append(key)

        for key in keys_to_delete:
            del kwargs[key]
        self.__attrs_init__(*args, **kwargs)  # pyright: ignore[reportAttributeAccessIssue]
```

```python
        params = {
            "model": self.model,
            "temperature": self.temperature,
            "user": self.user,
            "seed": self.seed,
            **self._params,
        }
```

Would allow for:
```
OpenAiChatPromptDriver(
        model="gpt-4o",
        top_k=0,
)
```

**Pull request closed without merging**